### PR TITLE
Request management

### PR DESCRIPTION
**Pull request by martyniukroman**

### CODE REVIEW

File: InternetHospital.BusinessLogic/Models/FeedbackViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace InternetHospital.BusinessLogic.Models
+{
+    public class FeedbackViewModel
+    {
+        public int Id {get;set;}
+        public string Text { get; set; }
+        public int FeedbackTypeId { set; get; }
+        public string FeedbackTypeName { set; get; }
+        public DateTime dateTime { set; get; }
```
Comment (TREN-D) [inline InternetHospital.BusinessLogic/Models/FeedbackViewModel.cs:11]: Rename it

File: InternetHospital.BusinessLogic/Services/FeedBackService.cs [hunk 2]
```diff
@@ -50,5 +49,96 @@ public IEnumerable<FeedBackType> GetFeedBackTypes()
...
+            List<UserModel> involvedUsers = new List<UserModel>();
+
+            foreach (var item in localFeedbacks)
+            {
+                involvedUsers.AddRange(_context.Users
+                    .Where(x => x.Id == item.UserId)
+                    .Select(x => Mapper.Map<User, UserModel>(x))
+                    );
+            }
+            if (involvedUsers != null)
```
Comment (TREN-D) [inline InternetHospital.BusinessLogic/Services/FeedBackService.cs]: return involvedUsers

File: InternetHospital.BusinessLogic/Services/FeedBackService.cs [hunk 3]
```diff
@@ -50,5 +49,96 @@ public IEnumerable<FeedBackType> GetFeedBackTypes()
...
+            }
+            
+            if (queryParameters.SearchByName != null)
+            {
+                var toLowerSearchParameter = queryParameters.SearchByName.ToLower();
+                feedbackViews = feedbackViews.Where(x =>
+                x.Text.ToLower()
+                .Contains(toLowerSearchParameter))
+                .ToList();
+            }
```
Comment (TREN-D) [inline InternetHospital.BusinessLogic/Services/FeedBackService.cs:122]: space

File: InternetHospital.WebApi/Controllers/FeedBackController.cs [hunk 4]
```diff
@@ -44,9 +44,39 @@ public IActionResult CreateFeedBack([FromBody]FeedBackCreationModel feedBackCrea
...
+        public IActionResult GetFeedBackTypes()
         {
-            return _feedBackService.GetFeedBackTypes();
+            return Ok(_feedBackService.GetFeedBackTypes());
+        }
+
+        [HttpGet("getinvolvedusers")]
+        public IActionResult GedInvolvedUsers()
+        {
+            return _feedBackService.GetUsers() == null ? (IActionResult)BadRequest(null) : Ok(_feedBackService.GetUsers());
```
Comment (TREN-D) [inline InternetHospital.WebApi/Controllers/FeedBackController.cs]: remove null

File: InternetHospital.WebApi/Controllers/FeedBackController.cs [hunk 5]
```diff
@@ -44,9 +44,39 @@ public IActionResult CreateFeedBack([FromBody]FeedBackCreationModel feedBackCrea
...
+                });
+        }
+
+        [HttpPut("updatefeedback")]
+        public IActionResult UpdateFeedback([FromBody]FeedbackViewModel feedback) 
+        {
+
+            var reply = _feedBackService.UpdateFeedBack(feedback);
+
+            if (reply != null)
```
Comment (TREN-D) [inline InternetHospital.WebApi/Controllers/FeedBackController.cs]: why no ternar?

**Review by TREN-D — APPROVED** [review]

**Review by edward0802 — APPROVED** [review]

**Review by eugent482 — APPROVED** [review]

File: InternetHospital.BusinessLogic/Services/FeedBackService.cs [hunk 6]
```diff
@@ -1,36 +1,32 @@
...
 using Microsoft.Extensions.Logging;
 using System;
 using System.Collections.Generic;
 using System.Linq;
 
 namespace InternetHospital.BusinessLogic.Services
 {
-    public class FeedBackService : IFeedBackService
-    {
+    public class FeedBackService : IFeedBackService {
```
Comment (OleksandrAntoniuk) [inline InternetHospital.BusinessLogic/Services/FeedBackService.cs]: format brackets

File: InternetHospital.BusinessLogic/Services/FeedBackService.cs [hunk 7]
```diff
@@ -1,36 +1,32 @@
...
             _context = context;
         }
 
-        public FeedBack FeedBackCreate(FeedBackCreationModel model)
-        {
+        public FeedBack FeedBackCreate(FeedBackCreationModel model) {
             FeedbackValidator validations = new FeedbackValidator();
-            try
-            {
+            try {
```
Comment (OleksandrAntoniuk) [inline InternetHospital.BusinessLogic/Services/FeedBackService.cs]: and here

File: InternetHospital.BusinessLogic/Services/FeedBackService.cs [hunk 8]
```diff
@@ -39,16 +35,103 @@ public FeedBack FeedBackCreate(FeedBackCreationModel model)
                 _context.SaveChanges();
                 return feedback;
             }
-            catch (Exception exception)
-            {
-               _logger.LogCritical($"{nameof(FeedBackService)} failed with {exception.Message}", exception);
+            catch (Exception exception) {
```
Comment (OleksandrAntoniuk) [inline InternetHospital.BusinessLogic/Services/FeedBackService.cs]: everywhere please

File: InternetHospital.BusinessLogic/Services/FeedBackService.cs [hunk 9]
```diff
@@ -39,16 +35,103 @@ public FeedBack FeedBackCreate(FeedBackCreationModel model)
...
+
+            return involvedUsers;
+        }
+
+        public PageModel<List<FeedbackViewModel>> GetFeedbackViewModels(FeedbackSearchParams queryParameters) {
+            List<FeedBack> feedBacks = _context.FeedBacks
+                .Include(f => f.User)
+                .Include(f => f.Type)
+                .ToList();
+            IEnumerable<UserModel> users = GetUsers(feedBacks);
```
Comment (OleksandrAntoniuk) [inline InternetHospital.BusinessLogic/Services/FeedBackService.cs]: just var

File: InternetHospital.WebApi/Controllers/FeedBackController.cs [hunk 10]
```diff
@@ -44,9 +44,39 @@ public IActionResult CreateFeedBack([FromBody]FeedBackCreationModel feedBackCrea
...
```
Comment (OleksandrAntoniuk) [inline InternetHospital.WebApi/Controllers/FeedBackController.cs]: add brackets here or use ternary operator

File: InternetHospital.BusinessLogic/Services/FeedBackService.cs [hunk 11]
```diff
@@ -39,16 +35,103 @@ public FeedBack FeedBackCreate(FeedBackCreationModel model)
...
-               _logger.LogCritical($"{nameof(FeedBackService)} failed with {exception.Message}", exception);
+            catch (Exception exception) {
+                _logger.LogCritical($"{nameof(FeedBackService)} failed with {exception.Message}", exception);
                 return null;
             }
         }
 
-        public IEnumerable<FeedBackType> GetFeedBackTypes()
-        {
+        public IEnumerable<FeedBackType> GetFeedBackTypes() {
```
Comment (OleksandrAntoniuk) [inline InternetHospital.BusinessLogic/Services/FeedBackService.cs]: please add unit tests for this cases

**Review by OleksandrAntoniuk — CHANGES_REQUESTED** [review]

File: InternetHospital.BusinessLogic/Services/FeedBackService.cs [hunk 12]
```diff
@@ -39,16 +38,107 @@ public FeedBack FeedBackCreate(FeedBackCreationModel model)
...
             return _context.FeedBackTypes.ToList();
         }
+
+        public IEnumerable<UserModel> GetUsers(IEnumerable<FeedBack> feedBacks) {
```
Comment (OleksandrAntoniuk) [inline InternetHospital.BusinessLogic/Services/FeedBackService.cs]: brackets

File: InternetHospital.BusinessLogic/Services/FeedBackService.cs [hunk 13]
```diff
@@ -39,16 +38,107 @@ public FeedBack FeedBackCreate(FeedBackCreationModel model)
...
+            ICollection<UserModel> involvedUsers = new List<UserModel>();
+
+            var users = _context.Users;
+            foreach (var item in feedBacks) {
```
Comment (OleksandrAntoniuk) [inline InternetHospital.BusinessLogic/Services/FeedBackService.cs]: and here

File: InternetHospital.WebApi/Controllers/FeedBackController.cs [hunk 14]
```diff
@@ -44,9 +44,34 @@ public IActionResult CreateFeedBack([FromBody]FeedBackCreationModel feedBackCrea
...
+            return _feedBackService.GetUsers() == null ? (IActionResult)BadRequest() : Ok(_feedBackService.GetUsers());
+        }
+
+        [HttpGet("getviewfeedbacks")]
+        public IActionResult GetViewFeedbacks([FromQuery]FeedbackSearchParams queryParameters)
+        {
+            var result = _feedBackService.GetFeedbackViewModels(queryParameters);
+
+            return Ok(
+                new {
```
Comment (OleksandrAntoniuk) [inline InternetHospital.WebApi/Controllers/FeedBackController.cs]: here

File: InternetHospital.BusinessLogic/Models/FeedBackSearchParams.cs [hunk 15]
```diff
@@ -0,0 +1,17 @@
+﻿namespace InternetHospital.BusinessLogic.Models
+{
+    public class FeedbackSearchParams
+    {
+        private const int MAX_PAGE_COUNT = 50;
+        private const int DEFAULT_PAGE = 1;
+        public int Page { get; set; } = DEFAULT_PAGE;
```
Comment (OleksandrAntoniuk) [inline InternetHospital.BusinessLogic/Models/FeedBackSearchParams.cs:7]: why we can not init property here and delete redundant const?

File: InternetHospital.WebApi/Controllers/FeedBackController.cs [hunk 16]
```diff
@@ -35,11 +35,11 @@ public IActionResult CreateFeedBack([FromBody]FeedBackCreationModel feedBackCrea
             {
                 feedBackCreationModel.UserId = userid;
                 var feedback = _feedBackService.FeedBackCreate(feedBackCreationModel);
-                return feedback == null ? (IActionResult)BadRequest( new { message = "Not valid feedback" }) : Ok();
+                return feedback == null ? (IActionResult)BadRequest(new { message = "Not valid feedback" }) : Ok();
             }
             else
             {
-                return NotFound( new  { message = "Form not valid" });
+                return NotFound(new { message = "Form not valid" });
```
Comment (OleksandrAntoniuk) [inline InternetHospital.WebApi/Controllers/FeedBackController.cs]: form is not valid

File: InternetHospital.BusinessLogic/Services/FeedBackService.cs [hunk 17]
```diff
@@ -38,35 +39,39 @@ public FeedBack FeedBackCreate(FeedBackCreationModel model)
...
 
-        public IEnumerable<UserModel> GetUsers() 
+        public IEnumerable<UserModel> GetUsers()
         {
             var feedBacks = _context.FeedBacks;
             ICollection<UserModel> involvedUsers = new List<UserModel>();
 
             var users = _context.Users;
-            foreach (var item in feedBacks) 
+            foreach (var item in feedBacks)
```
Comment (OleksandrAntoniuk) [inline InternetHospital.BusinessLogic/Services/FeedBackService.cs]: feedbacks not feedBacks

File: InternetHospital.BusinessLogic/Services/FeedBackService.cs [hunk 18]
```diff
@@ -14,22 +14,23 @@
 
 namespace InternetHospital.BusinessLogic.Services
 {
-    public class FeedBackService : IFeedBackService
```
Comment (OleksandrAntoniuk) [inline InternetHospital.BusinessLogic/Services/FeedBackService.cs]: please rename to FeedbackService this class and methods with same names

**Review by OleksandrAntoniuk — APPROVED** [review]

**Review by OleksandrAntoniuk — APPROVED** [review]

**Pull request merged**